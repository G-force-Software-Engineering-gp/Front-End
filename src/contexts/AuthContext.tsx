import { createContext, useState, useEffect } from "react";
import jwt_decode from "jwt-decode";
import { useNavigate } from "react-router-dom";
import axios from "axios";
// import swal from "sweetalert";
// import axiosInstance from "../utils/axiosInstance";


export type AuthContextType = {
  user: any; // Replace with your user type
  setUser: React.Dispatch<React.SetStateAction<any>>;
  authTokens: any; // Replace with your auth tokens type
  setAuthTokens: React.Dispatch<React.SetStateAction<any>>;
  loginUser: (values: any) => Promise<void>;
  registerUser: (values: any) => Promise<void>;
  logoutUser: () => void;
};

const AuthContext = createContext<AuthContextType | null>(null);

export default AuthContext;
type AuthProviderProps = {
    children: React.ReactNode;
  };
  
  export const AuthProvider: React.FC<AuthProviderProps> = ({ children }) => {
  const [authTokens, setAuthTokens] = useState<any>(
    localStorage.getItem("authTokens")
      ? JSON.parse(localStorage.getItem("authTokens")!)
      : null
  );
  const [user, setUser] = useState<any>(
    localStorage.getItem("authTokens")
      ? jwt_decode(localStorage.getItem("authTokens")!)
      : null
  );
  const [loading, setloading] = useState(true);

  const Navigate = useNavigate();

  const loginUser = async (values: any) => {
    const response = await fetch(
        "https://amirmohammadkomijani.pythonanywhere.com/auth/jwt/create/",
        {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            email: values.email,
            password: values.password,
          }),
        }
      );
      const data = await response.json();
      console.log(data)
      if (response.status === 200) {
        setAuthTokens(data);
        setUser(jwt_decode(data.access));
        localStorage.setItem("authTokens", JSON.stringify(data));
        Navigate("/");
      } 
      else{
        console.log(data.error);
      }
  };
  //axios approach
        // await axios.post("https://amirmohammadkomijani.pythonanywhere.com/auth/jwt/create/",
        //     {
        //     email: values.email,
        //     password: values.password,
        //     },
        //     {
        //     headers: {
        //     'Content-Type': 'application/json',
        //     },
        //     })
        //     .then((response) => {
        //     const data = response.data;
        //     console.log(data);
        
        //     if (response.status === 200) {
        //         setAuthTokens(data);
        //         setUser(jwt_decode(data.access));
        //         localStorage.setItem('authTokens', JSON.stringify(data));
        //         Navigate('/');
        //     } else {
        //         console.log(data.error);
        //     }
        //     })
        //     .catch((error) => {
        //     console.error('API request failed:', error);
        //     });
        // }
  const registerUser = async (values: any) => {
      const reg_response = await fetch(
      "https://amirmohammadkomijani.pythonanywhere.com/auth/users/",
      {
          method: "POST",
          headers: {
          "Content-Type": "application/json",
          },
          body: JSON.stringify({
          first_name: values.firstName,
          last_name: values.lastName,
          email: values.email,
          username: values.username,
          password: values.password,
          }),
      });
      console.log(reg_response);
      const reg_data = await reg_response.json();
      if (reg_response.status === 201) {
        Navigate("/login");
      } else {
          console.log(reg_data.error);  
      }
  };

  //axios approach
        // await axios.post("https://amirmohammadkomijani.pythonanywhere.com/auth/users/",
        //     {
        //     first_name: values.firstName,
        //     last_name: values.lastName,
        //     email: values.email,
        //     username: values.username,
        //     password: values.password,
        //     },
        //     {
        //     headers: {
        //     'Content-Type': 'application/json',
        //     },
        //     })
        //     .then((reg_data) => {
        //         const data = reg_data.data;
        //         console.log(data);
        //         if (reg_data.status === 201) {
        //           Navigate('/login');
        //         } else {
        //           console.log(data.error);
        //         }
        //       })
        //       .catch((error) => {
        //         console.error('API request failed:', error);
        //       });
        // }

  const logoutUser = () => {
    setAuthTokens(null);
    setUser(null);
    localStorage.removeItem("authTokens");
    Navigate("/login");
  };

  const contextData: AuthContextType = {
    user,
    setUser,
    authTokens,
    setAuthTokens,
    loginUser,
    registerUser,
    logoutUser,
  };

  useEffect(() => {
    if (authTokens) {
        setUser(jwt_decode(authTokens.access));
      }
      setloading(false);
  }, [authTokens, loading]);

  return <AuthContext.Provider value={contextData}>{children}</AuthContext.Provider>;
};
