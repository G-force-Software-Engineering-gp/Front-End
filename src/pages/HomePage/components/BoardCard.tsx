import { Button } from '@/components/ui/button';
import { Card, CardHeader, CardTitle } from '@/components/ui/card';
import { StarIcon } from 'lucide-react';
import React from 'react';
import { useNavigate } from 'react-router-dom';

interface BoardCardProps {
  id: number;
  title: string;
  backgroundImage: any;
}

const BoardCard: React.FC<BoardCardProps> = ({ id, title, backgroundImage }) => {
  const navigate = useNavigate();
  return (
    <Card
      style={{ backgroundImage: `url(${backgroundImage})`, backgroundSize: 'cover' }}
      className=" h-36 cursor-pointer bg-slate-200 dark:bg-slate-900"
      onClick={() => navigate(`/board/${id}`)}
    >
      <CardHeader className="flex justify-between gap-4 space-y-0 p-4">
        <div className="flex justify-between">
          <div className="space-y-1">
            <CardTitle>{title}</CardTitle>
          </div>
          <Button variant="ghost">
            <StarIcon className="h-6 w-6" />
          </Button>
        </div>
      </CardHeader>
    </Card>
  );
};

export default BoardCard;
