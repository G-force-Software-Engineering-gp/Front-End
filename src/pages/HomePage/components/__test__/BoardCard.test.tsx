import { AuthProvider } from '@/contexts/AuthContext';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { MemoryRouter } from 'react-router-dom';
import BoardCard from '../BoardCard';

describe('BoradCard Component', () => {
  test('renders BoradCard component', () => {
    const mockProps = {
      id: 1,
      title: 'Test Board',
      backgroundImage: 'test-image-url',
      has_star: false,
    };
    render(
      <MemoryRouter>
        <AuthProvider>
          <BoardCard {...mockProps} />
        </AuthProvider>
      </MemoryRouter>
    );
  });
  // test('matches snapshot', () => {
  //   const mockProps = {
  //     id: 1,
  //     title: 'Test Board',
  //     backgroundImage: 'test-image-url',
  //     has_star: false,
  //   };
  //   const { asFragment } = render(
  //     <MemoryRouter>
  //       <AuthProvider>
  //         <BoardCard {...mockProps} />
  //       </AuthProvider>
  //     </MemoryRouter>
  //   );
  //   expect(asFragment()).toMatchSnapshot();
});
test('clicking on the card triggers navigation', () => {
  const mockProps = {
    id: 1,
    title: 'Test Board',
    backgroundImage: 'test-image-url',
    has_star: false,
  };
  render(
    <MemoryRouter>
      <AuthProvider>
        <BoardCard {...mockProps} />
      </AuthProvider>
    </MemoryRouter>
  );
  fireEvent.click(screen.getByText('Test Board'));
  // Add assertions based on the expected behavior after clicking the card
});
test('renders with correct star state', () => {
  const mockPropsWithStar = {
    id: 1,
    title: 'Test Board',
    backgroundImage: 'test-image-url',
    has_star: true,
  };
  const mockPropsWithoutStar = {
    id: 2,
    title: 'Another Board',
    backgroundImage: 'another-image-url',
    has_star: false,
  };

  const { getByTestId: getByTestIdWithStar } = render(
    <MemoryRouter>
      <AuthProvider>
        <BoardCard {...mockPropsWithStar} />
      </AuthProvider>
    </MemoryRouter>
  );
  const { getByTestId: getByTestIdWithoutStar } = render(
    <MemoryRouter>
      <AuthProvider>
        <BoardCard {...mockPropsWithoutStar} />
      </AuthProvider>
    </MemoryRouter>
  );

  const starIconWithStar = getByTestIdWithStar('star-icon1');
  const starIconWithoutStar = getByTestIdWithoutStar('star-icon2');

  expect(starIconWithStar).toBeInTheDocument();
  expect(starIconWithoutStar).toBeInTheDocument();
});

//   test('clicking on the star icon calls SetStarOrNot function', () => {
//     const mockProps = {
//       id: 1,
//       title: 'Test Board',
//       backgroundImage: 'test-image-url',
//       has_star: false,
//     };
//     const setStarOrNotMock = jest.fn();
//     jest.spyOn(React, 'useState').mockReturnValue([false, setStarOrNotMock]);

//     render(
//       <MemoryRouter>
//         <AuthProvider>
//           <BoardCard {...mockProps} />
//         </AuthProvider>
//       </MemoryRouter>
//     );
//     const starIcon = screen.getByTestId('star-icon1');
//     fireEvent.click(starIcon);

//     expect(setStarOrNotMock).toHaveBeenCalledWith(expect.any(Object));
//     // Add assertions based on the expected behavior after clicking the star icon
//   });
// });
