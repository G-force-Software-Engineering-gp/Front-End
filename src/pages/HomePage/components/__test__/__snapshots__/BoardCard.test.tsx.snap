// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoradCard Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="rounded-lg border text-card-foreground shadow-sm h-36 cursor-pointer bg-slate-200 dark:bg-slate-900"
    style="background-image: url(test-image-url); background-size: cover;"
  >
    <div
      class="flex-col flex justify-between gap-4 space-y-0 p-4"
    >
      <div
        class="flex justify-between"
      >
        <div
          class="space-y-1"
        >
          <h3
            class="text-2xl font-semibold leading-none tracking-tight"
          >
            Test Board
          </h3>
        </div>
        <button
          class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2"
        >
          <svg
            class="h-8 w-8"
            data-testid="star-icon2"
            fill="none"
            height="24"
            stroke="white"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="3"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polygon
              points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
